GN5 to GN4 proxy security using GN4's JWT Headers

davidblasby opened the pull request:
This is an experiment for passing security info from GN5 to GN4 via GN5's gateway proxy.

This PR includes two changes;
1. adding the "simple" and "complex" GN5->GN4 `cloud: gateway: mvc` filter
     + almost everyone will use the "simple"
    + requires GN4 to have JWT Headers configured (see below)
    + requires that GN4 is not accessible EXCEPT via the gateway - or you need to remove a header from incoming request
2. Minor changes to the OIDC so it works with github for users that don't have their email addresses public


## Securing GN5 to GN4 with signed JWT


A more secure way to secure send information to GN4 is to use signed JWT.  GN4 will verify they are correctly signed and prevent a malicious actor from faking a GN4 authentication token.

To set this up:

1. Generate an RSA private and public key
2. Create a JWK Set containing the public key
3. Configure GN5
4. Configure GN4

### Generate an RSA private and public key

```
openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 -keyout private.key -out certificate_pub.crt
```

This will generate:
1. `private.key` - use this in GN5
2. `certificate_pub.crt` -- use this in the next step for GN4

 ### Create a JWK Set containing the public key

The standard way of handling JWT signature validation is to use a JWK Set.  This is fairly easy to do.

1. Use [https://jwkset.com/generate](https://jwkset.com/generate) to convert your `certificate_pub.crt` to a JWK.
    a) copy the text from `certificate_pub.crt` into the first section - "`PEM encoded key or certificate`"
    b) choose a `Key ID` (you'll need this, below)
    c) `Key Algorithm` is "RSA256"
    d) `Key Use` is "Signature"
    e) press "Generate"
2. Put the result in file and save for GN4 configuration.
```
{
	"keys": [
		 ... TEXT FROM ABOVE ...
	]
}
```

NOTE: There are other online/offline tools that do this.

### Configure GN5

In `application.yml`:

```
  cloud:
    gateway:
      mvc:
        routes:
         ...
            filters:
              - addSimpleJwtGn4SecurityHeader=gn5.to.gn4.trusted.json.auth,file:///Users/db/delme/key/private.key,mykeyid
```

The parameters to the `addSimpleJwtGn4SecurityHeader` filter are:
a) name of the header to use (must be the same as the GN4 configuration)
b) URL to the private key
c) Name of the `Key ID` (from above)

### Configure GN4

Sent environment variables like this;

```
JWTHEADERS_ValidateTokenAudienceClaimValue=g4.from.g5.proxy
JWTHEADERS_ValidateTokenAudience=true
JWTHEADERS_ValidateTokenAudienceClaimName=aud
JWTHEADERS_ValidateTokenExpiry=true
JWTHEADERS_ValidateToken=true
JWTHEADERS_ValidateTokenSignatureURL=file:///Users/db/delme/jws.json
JWTHEADERS_UpdateProfile=false
JWTHEADERS_UpdateGroup=false
JWTHEADERS_RolesHeaderName=gn5.to.gn4.trusted.json.auth
JWTHEADERS_UserNameFormat=JWT
JWTHEADERS_JwtHeaderRoleSource=DB
JWTHEADERS_UserNameJsonPath=username
JWTHEADERS_ValidateTokenAgainstURL=false
JWTHEADERS_UserNameHeaderName=gn5.to.gn4.trusted.json.auth
```

#### Audience Validation

This audience is put into the JWT by GN5.  Its optional to validate it.

`JWTHEADERS_ValidateTokenAudienceClaimValue=g4.from.g5.proxy` 
`JWTHEADERS_ValidateTokenAudience=true`
`JWTHEADERS_ValidateTokenAudienceClaimName=aud` 

#### Token Expiry

`JWTHEADERS_ValidateTokenExpiry=true`

#### Validate Signature

This should be URL to the JWKSet you created, above.  
NOTE: see below if you are using a file instead of a https location 

`JWTHEADERS_ValidateTokenSignatureURL=file:///Users/db/delme/jws.json`

#### Other configuration

 
`JWTHEADERS_UserNameFormat=JWT` -- expect a JWT

`JWTHEADERS_UserNameJsonPath=username` -- json path (inside the payload) of the username
`JWTHEADERS_ValidateTokenAgainstURL=false` -- do not validate the token against the IDP
`JWTHEADERS_UserNameHeaderName=gn5.to.gn4.trusted.json.auth`- name of the Header for the JWT.  Must be the same as the GN5 configuration

`JWTHEADERS_UpdateProfile=false` - manage user permissions and groups in the DB
`JWTHEADERS_UpdateGroup=false`- manage user permissions and groups in the DB
`JWTHEADERS_JwtHeaderRoleSource=DB` - manage user permissions and groups in the DB

#### JWKSet in File or HTTP

The older implementation of JWT Headers isn't compatible with file-base URLs for the JWKSet.  This is fixed, but it will take a while for this to become "official" through the build chains.

Use `python` to serve your JWKSet file;

1. make a directory and put the JWKSet json file in it
2. run:
` python3 -m http.server 9000`
3. Use "http://localhost:9000/jwksets.json" 

See the python documentation for security related to doing this.


## Simple mode
### GN5 setup

In `application.yml` (cloud: gateway: mvc), add the "filters" (at the bottom):
```yaml
  cloud:
    gateway:
      mvc:
        routes:
          - id: geonetwork_proxy_redirect
            uri: ${geonetwork.openapi.url}
            predicates:
              - Path=/geonetwork/proxy
            filters:
              - RewritePath=/geonetwork/(?<url>.*), /api/$\{url}
          - id: geonetwork_route
            uri: ${geonetwork.core.url}
            predicates:
              - Path=/geonetwork/**
            filters:
              - addSimpleGn4SecurityHeader=gn5.to.gn4.trusted.json.auth
```
`gn5.to.gn4.trusted.json.auth` is the name of the header that will be attached to the requests so auth is recognized by GN4's JWT Headers Security module (see below).

This header will, typically, look like

```json
{
      "username":"admin"
}      
```


### GN4 setup

1. You will need the JWT Headers security model (currently in PR)
5. Set it up with the environment variables, below
6. Run with JWT Headers enabled `-Dgeonetwork.security.type=jwt-headers`
7. GN4 must **NOT** be accessible other than via the Gateway.  It is trusting headers so they must be removed.
    * if GN4 is available, you **MUST** remove the `gn5.to.gn4.trusted.json.auth` header from incomming requests
    * See your web environment documentation for how to do this
    
```sh
JWTHEADERS_UserNameHeaderName=gn5.to.gn4.trusted.json.auth
JWTHEADERS_UserNameFormat=JSON
JWTHEADERS_UserNameJsonPath=username
JWTHEADERS_JwtHeaderRoleSource=DB
JWTHEADERS_UpdateProfile=false
JWTHEADERS_UpdateGroup=false
```    

What these environment variables do:

*  `JWTHEADERS_UserNameHeaderFormat`  - name of the header to use (should be the same as your GN5 `application.yml` filter config)
* `JWTHEADERS_UserNameFormat` - should be JSON (the header is a JSON string)
* `JWTHEADERS_UserNameJsonPath` - where in the JSON is the username
* `JWTHEADERS_JwtHeaderRoleSource` - this should be `DB` (GN DB will manage user-profile-groups)
* `JWTHEADERS_UpdateProfile` -  JWT Headers should **NOT** update the DB profiles (user allowed to edit in GUI)
* `JWTHEADERS_UpdateGroup` -  JWT Headers should **NOT** update the DB user-groups (user allowed to edit in GUI)


See the JWT Headers documentation for more info.


### GitHub OIDC

To use with GitHub OIDC:

1. you'll need to create an app in github - you'll need to get the clientid and generate a clientsecret.
4. in `application.yml`, 

```yaml
  security:
    oauth2:
      client:
        registration:
          github:
            clientId:  ...
            clientSecret: ...
```
5. Have the user login to GN5 using the "GitHub" login on the homepage
6. Login as admin and edit the user so they have the appropriate permissions and info

NOTE: GN user name will either be the github email address (if its PUBLIC), or the github username.
 
## Complex Version


For most people, the Simple version is what you should use.  Only use the complex version in situation where the DB is not the authority on user rights.

1. In GN5 application.yml, use `addGn4SecurityHeader` as the filtername (no "simple")
3. in GN4:

```sh
JWTHEADERS_UserNameHeaderName=gn5.to.gn4.trusted.json.auth
JWTHEADERS_UserNameFormat=JSON
JWTHEADERS_UserNameJsonPath=username
JWTHEADERS_JwtHeaderRoleSource=JSON
JWTHEADERS_RolesHeaderName=gn5.to.gn4.trusted.json.auth
JWTHEADERS_RolesJsonPath=userGroupProfile
```    

The typical JSON will look like this:
```json
{
    "username":"dave",
    "userGroupProfile":  [
            "sample:RegisteredUser",
            "sample:Editor",
            "davegroup:RegisteredUser",
            "Editor"
    ]
} 
```

david-blasby committed:
GN5 to GN4 proxy security using GN4's JWT Headers

david-blasby committed:
update for -Dqa

david-blasby committed:
suppress incorrect findbug errors

fxprunayre commented on src/apps/geonetwork/src/main/java/org/geonetwork/gn4forwarding/SimpleGn4SecurityHeaderAppender.java:
```diff
@@ -0,0 +1,140 @@
+
+        var user = getUser(request);
+        if (user == null) {
+            return null;
```
```suggestion
            return changedRequest.build();
```

we must return a request here. Not 100% sure if we return this or the incoming request?

davidblasby replied:
good catch!

fxprunayre commented on src/apps/geonetwork/src/main/java/org/geonetwork/gn4forwarding/SimpleGn4SecurityHeaderAppender.java:
```diff
@@ -0,0 +1,140 @@
+
+        var username = user.getUsername();
+        if (username == null) {
+            return null;
```
```suggestion
            return changedRequest.build();
```

Same here

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=77) <br/>All committers have signed the CLA.

david-blasby committed:
add support for signed JWT + simplied the spring integration + added abstract class

davidblasby commented:
ok,
1. added support for signed JWT - see above for documentation
2. refactored so there is an abstract parent class for the JWT/JSON output
3. some other simplifications

I haven't done any testing... yet

david-blasby committed:
minor format changes?

david-blasby committed:
add test cases.  note - enables testing in pom.xml

davidblasby commented:
I've added test cases.  However, I noticed that tests were disables (skipped) in the `app/geonetwork/pom.xml`.

This is likely because most of the test are integration tests and pull docker images and are slow.

However, this disables ALL tests, so I had to turn it on to get my unit test to run.

Those docker-based tests should probably be put in integration tests.

@josegar74 @fxprunayre

davidblasby commented on src/apps/geonetwork/src/main/java/org/geonetwork/gn4forwarding/AbstractGn4SecurityHeaderAppender.java:
```diff
@@ -0,0 +1,179 @@
+ */
+public abstract class AbstractGn4SecurityHeaderAppender {
+
+    protected final org.slf4j.Logger logger = LoggerFactory.getLogger(AbstractGn4SecurityHeaderAppender.class);
```
changed

david-blasby committed:
turn off integration tests with @Disabled

david-blasby committed:
jwt headers documentation

davidblasby commented:
added documentation (see `docs/manual`)

david-blasby committed:
changes for -Dqa

david-blasby committed:
update documentation with github login

david-blasby committed:
minor doc changes

fxprunayre commented on src/apps/geonetwork/src/main/java/org/geonetwork/gn4forwarding/SimpleGn4SecurityHeaderAppender.java:
```diff
@@ -0,0 +1,76 @@
+     * @return for spring to identify in application.yaml for the gateway (filters)
+     */
+    @Shortcut
+    public static HandlerFilterFunction<ServerResponse, ServerResponse> addSimpleGn4SecurityHeader2(String headerName) {
```
```suggestion
    public static HandlerFilterFunction<ServerResponse, ServerResponse> addSimpleGn4SecurityHeader(String headerName) {
```

fxprunayre commented on src/apps/geonetwork/src/main/java/org/geonetwork/gn4forwarding/SimpleJwtGn4SecurityHeaderAppender.java:
```diff
@@ -0,0 +1,85 @@
+     * @return for spring to identify in application.yaml for the gateway (filters)
+     */
+    @Shortcut
+    public static HandlerFilterFunction<ServerResponse, ServerResponse> addSimpleJwtGn4SecurityHeader2(
```
```suggestion
    public static HandlerFilterFunction<ServerResponse, ServerResponse> addSimpleJwtGn4SecurityHeader(
```

fxprunayre committed:
Gateway / GN4 route / Enable security by default.

Will work fine with https://github.com/geonetwork/core-geonetwork/pull/8559

fxprunayre commented on docs/manual/docs/GN4-Integration/auth-jwt.md:
```diff
@@ -0,0 +1,135 @@
+
+This is the meaning of the other configuration options:
+
+`JWTHEADERS_UserNameFormat=JWT` -- expect indentity information to be in a JWT
```
```suggestion
`JWTHEADERS_UserNameFormat=JWT` -- expect identity information to be in a JWT
```

davidblasby committed:
Update docs/manual/docs/GN4-Integration/auth-jwt.md

Co-authored-by: François Prunayre <fx.prunayre@gmail.com>

davidblasby committed:
Update src/apps/geonetwork/src/main/java/org/geonetwork/gn4forwarding/SimpleGn4SecurityHeaderAppender.java

Co-authored-by: François Prunayre <fx.prunayre@gmail.com>

davidblasby committed:
Update src/apps/geonetwork/src/main/java/org/geonetwork/gn4forwarding/SimpleJwtGn4SecurityHeaderAppender.java

Co-authored-by: François Prunayre <fx.prunayre@gmail.com>

davidblasby commented:
@fxprunayre  - thanks for review!

david-blasby committed:
add doc warnings for issues with JWT config

david-blasby committed:
removed local config

fxprunayre commented on config/application.yml:
```diff
@@ -100,6 +100,8 @@ spring:
             predicates:
               - Path=/geonetwork/**
+            filters:
+              #- addSimpleGn4SecurityHeader=gn5.to.gn4.trusted.json.auth
```
@davidblasby we should enable simple mode by default no?

```suggestion
              - addSimpleGn4SecurityHeader=gn5.to.gn4.trusted.json.auth
```

davidblasby committed:
Update config/application.yml

Co-authored-by: François Prunayre <fx.prunayre@gmail.com>

davidblasby commented:
@fxprunayre - yes, I think you're correct!  It is, however, a potential security issue...

fxprunayre reviewed (approved):
Tested with signed or not JWT. For the doc, we probably need to check how to organize it but we can do in another PR.
